src/QMCWaveFunctions/WaveFunctionFactory.cpp

```diff
@@ -158,7 +158,16 @@ std::unique_ptr<TrialWaveFunction> WaveFunctionFactory::buildTWF(xmlNodePtr cur,
   if (!vp_file_to_load.empty())
   {
     app_log() << "  Reading variational parameters from " << vp_file_to_load << std::endl;
-    dummy.readFromHDF(vp_file_to_load);
+    hdf_archive hin;
+    int major_version = OptimizableObject::openHDFToRead(vp_file_to_load, hin);
+    if (major_version == 1)
+      dummy.readFromHDF(vp_file_to_load);
+    else
+    {
+      UniqueOptObjRefs opt_obj_refs = targetPsi->extractOptimizableObjectRefs();
+      for (auto opt_obj : opt_obj_refs)
+        opt_obj.get().readVariationalParameters(hin, dummy);
+    }
```

Comment:
Versioning data files is pretty important.   Three digits is the standard for semantic versioning, and we ought to have a really good reason to use a different versioning scheme.

Only the major version was returned because that's all that is needed for now.  If future implementations need more of the version, the code can be changed.

It sounds like the design you're asking for is #4517, but with the name of {read,save}ExtraParameters changed to drop the "extra" part of the name.

The big difference between that PR and this one is how the implementation default is handled for OptimizableObjects that don't handle reading/writing their own variational parameters.

Reply:
> It sounds like the design you're asking for is https://github.com/QMCPACK/qmcpack/pull/4517, but with the name of {read,save}ExtraParameters changed to drop the "extra" part of the name.

yes. Sort of. The reason to drop "extra" is to have per OptimizableObject I/O not just doing the "extra" parameters but also the variational parameters this OptimizableObject object owns. So we can have the global buffer I/O as only a secondary method and rely on more precise per component I/O as the primary method.